terraform/eval_for_each.go

```diff
@@ -0,0 +1,73 @@
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/terraform/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// evaluateResourceForEachExpression interprets a "for_each" argument on a resource.
+//
+// Returns a cty.Value map, and diagnostics if necessary. It will return nil if
+// the expression is nil, and is used to distinguish between an unset for_each and an
+// empty map
+func evaluateResourceForEachExpression(expr hcl.Expression, ctx EvalContext) (forEach map[string]cty.Value, diags tfdiags.Diagnostics) {
+	if expr == nil {
+		return nil, nil
+	}
+
+	forEachVal, forEachDiags := ctx.EvaluateExpr(expr, cty.DynamicPseudoType, nil)
+	diags = diags.Append(forEachDiags)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	// No-op for dynamic types, so that these pass validation, but are then populated at apply
+	if forEachVal.Type() == cty.DynamicPseudoType {
```

Comment:
When I was playing around with the latest `for_each` implementation I accidentally triggered a funny case where I got the following error in a situation where I _had_ `for_each` set:

```
Error: Reference to "each" in context without for_each

  on ../../main.tf line 113, in resource "dns_ns_record_set" "this":
 113:   nameservers = each.value.records

The "each" object can be used only in "resource" blocks, and only when the
for_each" argument is set.
```

After digging around a little, I suspect the block I attached this comment to is what led there... I managed to get into a situation where the `for_each` value was `cty.DynamicVal` by using a computed resource attribute as part of my `for_each` expression, which this codepath then treated as if it was not set at all.

In the equivalent situation for `count` we return a different error here:

```
Error: Invalid count argument

  on simple.tf line 73, in resource "null_resource" "foo":
  73:   count = uuid()

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.
```

We have #4149 tracking the idea of permitting unknown values in `count` (along with a potential solution), so I wonder if we should follow suit for `for_each` here and make this be a similar error for now (including the guidance about using `-target`), and then plan to treat `for_each` the same way as `count` when we design a final solution for #4149 in the future.

I lucked out here in that I created a situation where `cty.DynamicVal` was the result, but when working with resource attributes that are unknown during plan it's also possible that we'll see things like `cty.UnknownVal(cty.Set(cty.String))` here, if e.g. an AWS `availability_zones` set attribute can't be populated until apply, so to cover that I'd recommend using `!forEachVal.IsKnown()` for this case, which _includes_ `cty.DynamicVal` because it is equivalent to `cty.UnknownVal(cty.DynamicPseudoType)`, but also covers unknown values with known types.